pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/TransformPipeline.java

```diff
@@ -0,0 +1,149 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.data.manager.realtime;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import org.apache.pinot.segment.local.recordtransformer.ComplexTypeTransformer;
+import org.apache.pinot.segment.local.recordtransformer.CompositeTransformer;
+import org.apache.pinot.segment.local.recordtransformer.RecordTransformer;
+import org.apache.pinot.segment.local.utils.IngestionUtils;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+/**
+ * The class for transforming validating GenericRow data against table schema and table config.
+ * It is used mainly but not limited by RealTimeDataManager for each row that is going to be indexed into Pinot.
+ */
+public class TransformPipeline {
+  private final RecordTransformer _recordTransformer;
+  private final ComplexTypeTransformer _complexTypeTransformer;
+
+  /**
+   * Constructing a transform pipeline based on TableConfig and table schema.
+   * @param tableConfig the config for the table
+   * @param schema the table schema
+   */
+  public TransformPipeline(TableConfig tableConfig, Schema schema) {
+    // Create record transformer
+    _recordTransformer = CompositeTransformer.getDefaultTransformer(tableConfig, schema);
+
+    // Create complex type transformer
+    _complexTypeTransformer = ComplexTypeTransformer.getComplexTypeTransformer(tableConfig);
+  }
+
+  /**
+   * Process and validate the decoded row against schema.
+   * @param decodedRow the row data to pass in
+   * @param reusedResult the reused result so we can reduce objects created for each row
+   * @return both processed rows and failed rows in a struct.
+   * @throws TransformException when data has issues like schema validation. Fetch the partialResult from Exception
+   */
+  public Result processRow(GenericRow decodedRow, Result reusedResult) throws TransformException {
+    // to keep track and add to "failedRows" when exception happens
+    if (reusedResult == null) {
```

Comment:
`reusedResult` should never be `null`.
Going over the code again, I think we can just make the return `void`, and always ask the caller to pass in the result buffer. This way we don't need to explicitly try-catch the exception but let the caller decide how to handle the exception with the intermediate result in the result buffer. It is more flexible, and we can achieve different exception handling logic on the caller side instead of fixing it in this class.

Reply:
I like the idea. Thanks.